pybossa/jobs.py

```diff
@@ -1297,6 +1319,41 @@ def get_management_dashboard_stats(user_email):
     send_mail(mail_dict)
 
 
+def check_and_send_project_progress(project_id, conn=None):
+    from pybossa.cache.helpers import n_available_tasks
+    import pybossa.cache.users as cached_users
+    from pybossa.core import project_repo
+    from sqlalchemy.sql import text
+
+    project = project_repo.get(project_id)
+    if not project:
+        return
+
+    reminder = project.info.get('progress_reminder', {})
+    target_remaining = reminder.get("target_remaining")
+    email_already_sent = reminder.get("sent") or False
+    if target_remaining is None:
+        return
+
+    n_available_tasks = n_available_tasks(project.id)
+    if n_available_tasks > target_remaining:
+        reminder['sent'] = False
+    elif not email_already_sent:
+        # progress reached threshold and email not sent yet, send email
+        email_addr = [cached_users.get_user_email(user_id)
+                        for user_id in project.owners_ids]
+        info = dict(project_name=project.name,
+                    n_available_tasks=n_available_tasks)
+        notify_project_progress(info, email_addr)
+        reminder['sent'] = True
+
+    project.info['progress_reminder'] = reminder
+    if conn is not None:
```

Comment:
i think project_repo can be used to update reminder. Also, i think update is only needed when
```
(n_available_tasks > target_remaining and reminder['sent'] == True) or
(n_available_tasks < target_remaining and reminder['sent'] == False)
```
whereas in current case, update will occur with each update and at time with same value of reminder['sent']

Reply:
My intention was to use project_repo to update reminder too, but it didn't work. I assume because `check_and_send_project_progress` can be called in `event_listener`, see https://github.com/bloomberg/pybossa/blob/ad43604b08ae864f6da4d22997666eaecda25265/pybossa/model/event_listeners.py#L145 and https://github.com/bloomberg/pybossa/blob/ad43604b08ae864f6da4d22997666eaecda25265/pybossa/model/event_listeners.py#L250. 

My assumption is the listeners are under "connection" already, if we create another connection to database (inside project_repo.save()), will cause concurrency issue, which is prevented by sqlalchemy.

Reply:
yes, we need to update only when
```if (n_available_tasks > target_remaining and reminder['sent'] == True) or   
(n_available_tasks < target_remaining and reminder['sent'] == False)
```
will update code to reduce unnecessary connection with database.

Reply:
Thanks. Logging around this are would be helpful for troubleshooting issues when needed.

Reply:
right, will do